lib/FusionInventory/Agent/Tools/Hardware.pm

```diff
@@ -540,18 +559,56 @@ sub _getMacAddress {
     # fallback on ports addresses (IF-MIB::ifPhysAddress) if unique
     my $addresses_oid = ".1.3.6.1.2.1.2.2.1.6";
     my $addresses = $snmp->walk($addresses_oid);
+
+    # filter keeping mac with a defined and associated ip otherwise takes all macs
+    my $ips = $snmp->walk('.1.3.6.1.2.1.4.20.1.2');
+    my @macs = ();
+    @macs = grep { defined } map { $addresses->{$_} } values %{$ips}
+        if (keys(%{$ips}));
+    @macs = grep { defined } values %{$addresses}
+        unless @macs;
+
     my @addresses =
         uniq
+        grep { /^$mac_address_pattern$/ }
         grep { $_ ne '00:00:00:00:00:00' }
         grep { $_ }
         map  { _getCanonicalMacAddress($_) }
-        values %{$addresses};
+        @macs;
+
+    if (@addresses) {
+        return $addresses[0] if @addresses == 1;
 
-    return $addresses[0] if @addresses && @addresses == 1;
+        # Compute mac addresses as number and sort them
+        my %macs = map { $_ => _numericMac($_) } @addresses;
```

Comment:
Is this really need to convert those mac addresses before sorting them ? If the goal is just to retrieve a constant identifying value for each scan, not necessarily the actual lower value, a simple alphanumeric sort may achieve the same result.